Utils/src/L1NonPrefiringProbProducer.cc

```diff
@@ -0,0 +1,348 @@
+   iEvent.put( std::move(NonPrefiringProbMuonUp), "NonPrefiringProbMuonUp" );
+   iEvent.put( std::move(NonPrefiringProbMuonDown), "NonPrefiringProbMuonDown" );
+}
+
+std::array<double,3> L1NonPrefiringProbProducer::getNonPrefiringRateECAL( double eta, double pt, TH2F * h_prefmap) const {
+   if(h_prefmap==0) return {{0.,0.,0.}};
+   
+   //Check pt is not above map overflow
+   int nbinsy = h_prefmap->GetNbinsY();
+   double maxy = h_prefmap->GetYaxis()->GetBinLowEdge(nbinsy+1);
+   if(pt>=maxy) pt = maxy-0.01;
+   int thebin = h_prefmap->FindBin(eta,pt);
+   
+   double prefrate = h_prefmap->GetBinContent(thebin);
+   double preferr = h_prefmap->GetBinError(thebin);
+   double prefrate_up = std::min(std::max(prefrate + preferr, (1. + prefiringRateSystUncECAL_)*prefrate),1.);
+   double prefrate_dn = std::max(std::min(prefrate - preferr, (1. - prefiringRateSystUncECAL_)*prefrate),0.);
+   
+   return {{1.-prefrate, 1.-prefrate_up, 1.-prefrate_dn}};
+}
+
+std::array<double,3> L1NonPrefiringProbProducer::getNonPrefiringRateMuon( double eta, double phi, double pt) const {
+
+   double abseta = abs(eta);
+   double prefrate = 0.0; double preferr = 0.0;
+   if ((dataeramuon_.find("2016") != std::string::npos) && (eta > 1.24 && eta < 1.6) && (phi > 2.44346 && phi < 2.79253)){
+      prefrate = h_prefparam_muon_hotspot->Eval(pt);
+      preferr = h_prefparam_muon_hotspot->GetParError(2); 
+   } else if (abseta < 0.2) {
+      prefrate = h_prefparam_muon_0p00to0p20->Eval(pt);
+      preferr = h_prefparam_muon_0p00to0p20->GetParError(2);
+   } else if (abseta < 0.3) {
+      prefrate = h_prefparam_muon_0p20to0p30->Eval(pt);
+      preferr = h_prefparam_muon_0p20to0p30->GetParError(2);
+   } else if (abseta < 0.55) {
+      prefrate = h_prefparam_muon_0p30to0p55->Eval(pt);
+      preferr = h_prefparam_muon_0p30to0p55->GetParError(2);
+   } else if (abseta < 0.83) {
+      prefrate = h_prefparam_muon_0p55to0p83->Eval(pt);
+      preferr = h_prefparam_muon_0p55to0p83->GetParError(2);
+   } else if (abseta < 1.24) {
+      prefrate = h_prefparam_muon_0p83to1p24->Eval(pt);
+      preferr = h_prefparam_muon_0p83to1p24->GetParError(2);
+   } else if (abseta < 1.4) {
+      prefrate = h_prefparam_muon_1p24to1p40->Eval(pt);
+      preferr = h_prefparam_muon_1p24to1p40->GetParError(2);
+   } else if (abseta < 1.6) {
+      prefrate = h_prefparam_muon_1p40to1p60->Eval(pt);
+      preferr = h_prefparam_muon_1p40to1p60->GetParError(2);
+   } else if (abseta < 1.8) {
+      prefrate = h_prefparam_muon_1p60to1p80->Eval(pt);
+      preferr = h_prefparam_muon_1p60to1p80->GetParError(2);
+   } else if (abseta < 2.1) {
+      prefrate = h_prefparam_muon_1p80to2p10->Eval(pt);
+      preferr = h_prefparam_muon_1p80to2p10->GetParError(2);
+   } else if (abseta < 2.25) {
+      prefrate = h_prefparam_muon_2p10to2p25->Eval(pt);
+      preferr = h_prefparam_muon_2p10to2p25->GetParError(2);
+   } else if (abseta < 2.4) {
```

Comment:
this if/else if structure is essentially a linear search (O(N)) over bins. instead, the search can be done in O(logN) by making an `std::map<double,TF1*>`, where the keys are the bin upper boundaries. Then use the `std::map` accessor `upper_bound()` to get the right bin.

Reply:
Addressed in e9088f6